core/cap-0063.md

```diff
@@ -0,0 +1,496 @@
+  - Maximum number of evicted Soroban and classic entry read operations per ledger
+- `ledgerMaxReadEvictedAndClassicBytes`
+  - Maximum number of bytes of evicted Soroban and classic state that can be read per ledger
+- `txMaxReadEvictedAndClassicEntries`
+  - Maximum number of evicted Soroban and classic entry read operations per transaction
+- `txMaxReadEvictedAndClassicBytes`
+  - Maximum number of bytes of evicted Soroban and classic state that can be read per transaction
+- `feeReadArchiveAndClassicEntry`
+  - Fee per evicted Soroban/classic entry read
+- `feeReadArchiveAndClassic1KB`
+  - Fee for reading 1 KB of evicted Soroban/classic state
+    Maximum number of bytes of evicted Soroban and classic state that can be
+
+Note: write related values, including BucketList target size and fee calculation, have not changed semantically. Both
+Soroban and classic data share the same write fees.
+
+#### Initial Network Config Settings
+
+Initially, all evicted/classic limits and fees should match exactly the limits and fees of reads today. Limits must not
+decrease from current values in order to prevent bricking existing contracts. In the worst case, a contract can read
+up to `txMaxReadEntries - 2` classic entries today (all classic entries + wasm + instance). Current limits have been
+carefully measured assuming disk access only, so the addition of in-memory state should not negatively affect the network
+in any way should we maintain these values via evicted/classic limits.
+
+For fees, it would be beneficial for live Soroban state to be cheaper than classic/evicted state. However, the cost of
+classic/evicted state should not increase. From a network health perspective, in-memory reads should be preferred over
+disk access, so providing a discount incentives healthy contract behavior.
+
+Additionally, it is beneficial to encourage developers to use preflight regularly in preparation for full state archival.
+Should fees for live Soroban state be the same as in-memory state, developers might elect to bypass preflight and construct
+a "one shot" transaction. This would look like a transaction that only uses evicted/classic state and marks all Soroban entries
+as evicted via the bitmap. While this would work currently, in the future when full State Archival is introduced, this might
+break dApps, ad preflight would be required for proof generation and these "one shot" transactions would no longer be valid.
+By maintaining smaller fees for live Soroban state, we encourage developers to preflight Soroban transactions for lower fees
+in order to ensure ecosystem projects do not break with the addition of archival proofs.
+
+TODO: Exact fee and limit calculations
+
+#### `evictedSorobanEntries` BitMap
+
+For the purposes of block creation, the `LedgerKey` alone is sufficient to distinguish between classic and Soroban data.
+However, a disk read would be required prior to tx application in order to determine if a Soroban entry was live or evicted.
```

Comment:
is the disk read strictly necessary? Couldn't you do an in-memory lookup and the absence of data would indicate that is either evicted or non-existent?

Reply:
The issue is, if something is absent from the cache, it could be archived, or it could just be a new entry. You could imagine a tx that reads the maximum amount of classic state, doesn't restore anything, and write a new soroban entry. This tx should succeed, but when you check the cache for the new entry key, it won't exist. If you conservatively assume non-existent entries are always archived, then you filter out legitimate TXs. If you assume every cache miss is just a non-existent entry, then you will populate the block with TXs guaranteed to fail because they access too much Soroban state. This would limit our ability to construct optimal blocks.

Reply:
I thought writing a new entry would still require a disk read because you need to make sure that you are not attempting to create a new entry which already exists. And in order to check if the entry already exists you would need to do a disk lookup to see if the ledger key is in the hot archive bucketlist. I remember in another design doc you mentioned that core would maintain a binary fuse filter to make the duplicate ledger entry check more efficient.

Reply:
True, but checking if an entry doesn't exist in the database is a very different and much less expensive query then checking if something does exist in the database.

The new entry write process looks as follows:

```python
for each soroban key in readWrite:
    # Could either be archived entry or new entry
    if key not in cache:
        # Check in memory bloom filter for Hot Archive for key
        if hotArchiveBloomFilter.contains(key):
            entry = hotArchive.loadFromDisk(key)
            # Entry does not actually exist, filter false positive
            if not entry:
                continue
            # Entry does exist in archive, fail TX
            else:
                failTx
```

If a new entry actually does exist in the Hot Archive, we do a max of one disk read that the resources didn't pay for, then fail the TX. Note this happens at apply time where we've already charged fees, not a tx queue submission time. There may also be a bloom miss where a non-existent key requires a disk read, but this is rare enough to not be factor. TLDR: In the apply path, we make at most up to 1 non metered disk read and charge fees.

The key point with new entry creation is we assume no entries are archived so we can efficiently leverage the bloom filter and exit early at the first disk read. However, in your proposal we can't exit early at the first entry that is actually archived. Rather, we need to mark every key as non-archived vs archived. So for each archived entry, we need to do a disk read. The bloom filter is guaranteed to return positive for each archived entry, but you still need to do a disk read to see if it's a false positive or not. Additionally, these reads need to happen during tx flooding before we actually charge fees, which is a huge DOS angle.

Note that the bloom filters I'm talking about here are not the bloom filters in the state archival cap. These are local bloom filters used by validator DBs and are not deterministic between validators.